User experience workflow in OCP CoreOS Layering model and phased approach

sinnykumari opened the pull request:
Since the last time, we have gained experience while working
on the prototype in the layering branch https://github.com/openshift/machine-config-operator/tree/layering
and discussing/demoing to different teams.

Goal is to have up-to-date information of the existing
enhancement and sketch how the user experience will look like
in the layering model.

**Note** There will be some additional update added to this PR by the authors of this enhancement when spikes mentioned in this doc has been completed.

mrunalp commented on enhancements/ocp-coreos-layering.md:
````diff
@@ -103,23 +109,117 @@ A future enhancement will help automate the above.  For example, we may support
+###### Rollout of Final Image to Node
+- Once the built final pool image has been tested by the user, they can initiate a rollout to a specific pool:
+   ```
+   $ oc rollout custombuild mcp/worker
````
What about `custom-os-build`?

mrunalp commented on enhancements/ocp-coreos-layering.md:
````diff
@@ -103,23 +109,117 @@ A future enhancement will help automate the above.  For example, we may support
+  ```
+  $ oc add upgrade —layeredsource -f
+  ```
+- Final image gets generated in-cluster once MCO upgrades.
````
Where will this build run?
Will it be a nested build or run podman/buildah on a node?

cheesesashimi replied:
I think a nested build would be better especially as we consider multiarch cases. Something I'm cogitating on is what the nested build should look like.

mrunalp commented on enhancements/ocp-coreos-layering.md:
````diff
@@ -103,23 +109,117 @@ A future enhancement will help automate the above.  For example, we may support
+- Rollout follows the existing MCO rollout paradigm - i.e. the images roll out to max unavailable nodes in a pool. This will cause reboots.
+- After rollout has finished, there is an easy way to see which builds are present in the pools:
+  ```
+  $ oc get custombuilds -n openshift-machine-config-operator
````
Can we add an API section detailing the CRD?

dhellmann commented on enhancements/ocp-coreos-layering.md:
```diff
@@ -103,23 +109,117 @@ A future enhancement will help automate the above.  For example, we may support
+
+###### Create Customer Base Image
+- First of all the user will create a Customer Base Image. This image can run on compatible OCP clusters such as HyperShift, Heterogeneous cluster, Single Node OpenShift.
+- Create a Dockerfile that uses compatible rhcos image and add the custom content (e.g. rpm, files) on top of it:
```
It would help if each step in this workflow was explicit about who the actor is (the user, a software component, etc.) It can be a bit tedious to write that way, but the results are much clearer for a reader to follow.

sinnykumari replied:
Thanks for the feedback, will make the actor explicit in second pass.

dhellmann commented on enhancements/ocp-coreos-layering.md:
```diff
@@ -103,23 +109,117 @@ A future enhancement will help automate the above.  For example, we may support
+    - **On compatibility check failed**: “failed to tag quay.io/examplecorp/custom-rhel-coreos:4.11: rhcos version 89.2423423 not compatible with OCP 4.11.5, expected: rhcos version 89.43920432”
+
+###### Create Final Pool Images
+When the Customer Base Image has been successfully tagged into cluster, a final image will be created for each pool.
```
The previous section talks about tagging images, but not tagging them "into a cluster". Is that something different?

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
````diff
@@ -103,23 +127,201 @@ A future enhancement will help automate the above.  For example, we may support
+###### Providing Customer Base Image during cluster upgrade
+- OS can be updated with Customer Base Image during a cluster upgrade. For this, the user will need to create the Customer Base Image using the incoming rhcos version and supply to upgrade so that default osimageurl is not used:
+  ```
+  $ oc add upgrade —layeredsource -f
````
Could this be a ConfigMap that has a reference to the version the cluster is updating to and an os image url that is built for that version?

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -58,18 +57,34 @@ This is the OpenShift integration of [ostree native containers](https://fedorapr
+#### Phase 0
+- The `machine-os-content` shipped as part of the release payload will change format to the new "native ostree-container" format, in which the OS content appears as any other OCI/Docker container. Spike: [MCO-293](https://issues.redhat.com/browse/MCO-293).
+  (In contrast today, the existing `machine-os-content` has an ostree repository inside a UBI image, which is hard to inspect and cannot be used for derived builds).  For more information, see [ostree-rs-ext](https://github.com/ostreedev/ostree-rs-ext/) and [CoreOS layering](https://github.com/coreos/enhancements/pull/7).
+- Additionally, the MCO will be modified to allow a cluster admin to override the `osImageURL` field on MachineConfigs to supply a new base image.
```
Is this with a minimal CRD?

sinnykumari replied:
We won't be adding any CRD for phase 0

mrunalp replied:
Can we add an example here?

sinnykumari commented:
With our last OCP CoreOS layering sync-up, we agreed to focus on Phase 0 implementation. Phase 0 is what we are focusing during 4.12. Remaining phases may change based on feedback and learning we will will have.
@mrunalp EP is updated with upgrade workflow during phase 0 and is ready for review.

sinnykumari commented:
`make image` doesn't report error while running locally to me. Will look further why linter check is failing in ci.

jmencak commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
s/Hypershift/HyperShift/

jmencak reviewed (commented):
Thank you for a good read.  Question.  Will there be a mechanism to stage/queue changes for the OS layer without rebooting and then have them ready when we finally reboot?  Thinking multiple operators wanting their own layer at cluster install time and reducing reboots.  Or do we intend to use the existing functionality of pausing Machine Config Pools?

sinnykumari commented:
> Will there be a mechanism to stage/queue changes for the OS layer without rebooting and then have them ready when we finally reboot? Thinking multiple operators wanting their own layer at cluster install time and reducing reboots. Or do we intend to use the existing functionality of pausing Machine Config Pools?

With layering we are not explicitly looking for solving reboot behavior of MCO, it should work the way it works today. That would be handled by separate RFE.

On a similar note, we might explore to use `rpm-ostree ex apply-live` experimental feature to apply some of os changes without reboot

mrunalp commented on enhancements/ocp-coreos-layering.md:
```diff
@@ -79,17 +76,26 @@ An OpenShift administrator at example.corp is happily using OpenShift 4 (with RH
-#### Kernel hotfix
+A few weeks later, Red Hat releases a kernel security update as part of a cluster image, which appears as a new image at registry.redhat.io/rhel-coreos-8.6:4.12. This triggers a rebuild on their central cluster, which is successful and pushed to the mirror registry.
+
+The cluster is configured to poll for changes to this image in the remote registry, and when that happens, the updated image will be automatically rolled out to all nodes - and note this will happen *without* an `oc adm upgrade`.  (Note: rollout will be configurable)
```
This assumes work beyond phase 0?

sinnykumari replied:
yes

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -62,11 +62,29 @@ This is the OpenShift integration of [ostree native containers](https://fedorapr
+
+#### Phase 1
+- Iterate on the mechanism established in Phase 0 by enabling rendered MachineConfigs to be layered into a final OS image before they are applied to a nodes’ disk by using the BuildController. (see [Create Final Pool Images](#create-final-pool-images)).
+- The BuildController will be reliant upon the OpenShift Image Builder. Alternatively, we could create our own custom solution (see: [How The Final Image Will Be Built](#how-the-final-image-will-be-built) for caveats).)
```
Does this cover day 0 support for drivers?

sinnykumari replied:
Not sure about details, leaving this in favor of phase 0 as these details may change.

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -272,13 +272,56 @@ We need to ensure that a cluster admins' customized image contains the same comp
-Related spikes:
+#### Where the image will live
 
+For [Phase 0](#phase-0) operation, the MCO doesn't have to worry about where the image is stored, since it's not doing any of the building. 
```
nit: since it's not responsible for building the images.

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -272,13 +272,56 @@ We need to ensure that a cluster admins' customized image contains the same comp
 
+For [Phase 0](#phase-0) operation, the MCO doesn't have to worry about where the image is stored, since it's not doing any of the building. 
+
+For [Phase 1](#phase-1) and [Phase 2](#phase-2), we may be performing in-cluster ["builds"](#how-the-final-image-will-be-built) and we would then need a place to keep the result and serve it up for at least:
```
nit: we may perform..

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -272,13 +272,56 @@ We need to ensure that a cluster admins' customized image contains the same comp
+This alternative registry could be:
+
+- a user-supplied external registry + pull/push credential ( strawman: quay.io )
+- an MCO-supplied minimal registry such that the MCO would be able to serve and store the images it produces
```
A customer deployed 3rd party registry such as artifactory, etc.

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -67,11 +67,12 @@ This is the OpenShift integration of [ostree native containers](https://fedorapr
-- In this phase, we can avoid performing a full image build. However, this approach has certain tradeoffs (see: [Buildless layering support](#buildless-layering-support)). 
-- Initial [preflight checks](#preflight-checks) are implemented. However, they will only be a warning and will not block image rollout since we will not (yet) have a way to bypass them.
+- This new format container will *also* be shipped alongside the toplevel release image with a matching version scheme.  For example, there will now be e.g `quay.io/openshift-release-dev/rhel-coreos-8:4.12.0-rc.5-x86_64`.
+- The MCO will be modified to allow a cluster admin to override the `osImageURL` field on MachineConfigs to supply a new base image that overrides the default (RHEL) CoreOS image.  The administrator can use the above image tags in a `Dockerfile` or other container build system.
```
Could we add an example here?

sinnykumari replied:
Not sure if there is much to add in example. It would be basic containerfile and MachineConfig containing osImageURL with custom os image url link. I will lean towards having these additional details in our user doc. Let me know if you still think we should adding here is useful.

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -531,7 +543,8 @@ Multiarch Spike: https://issues.redhat.com/browse/MCO-277
 #### Hypershift
 
-[Hypershift]https://github.com/openshift/hypershift has its own implementation of inplace upgrading for nodepools today. A special mode of the MCD is used in conjunction with Hypershift’s hosted control plane operator to facilitate this. We would definitely want feature parity in Hypershift eventually, as well as the benefits layering workflow would have for managing large amounts of hosted clust ...
+[Hypershift]https://github.com/openshift/hypershift has its own implementation of inplace upgrading for nodepools today. A special mode of the MCD is used in conjunction with Hypershift’s hosted control plane operator to facilitate this.
```
nit: inplace upgrades for..

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,725 @@
+
+## Summary
+
+Change RHEL CoreOS as shipped in OpenShift to be a "base image" that can be used as in layered container builds and then booted.  This will allow custom 3rd party agents delivered via RPMs installed in a container build.  The MCO will roll out and monitor these custom builds the same way it does for the "pristine" CoreOS image today.
```
nit: used in layered..

mrunalp replied:
The MCO will roll out and monitor these custom containers..
(once built it is just like any other other container)

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,725 @@
+### Goals
+
+- [ ] Administrators can add custom code alongside configuration via an opinionated, tested declarative build system and via e.g. `Dockerfile` as an unvalidated generic mechanism
+- [ ] The output is a container that can be pushed to a registry, inspected, and processed via security scanners. 
```
nit: consistency in periods at the end of each line in a section.

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,725 @@
+
+Note we will continue to support MachineConfig.  The key goal is to allow the user to override the rhel-coreos base image.
+
+#### Current default proposal: Configure via imagestream
```
This isn't the default anymore, so maybe we retitle the section?

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,725 @@
+
+This is currently prototyped in the https://github.com/openshift/machine-config-operator/tree/layering branch.  Here there’s an imagestream per MachineConfigPool, and the administrator can specify an image which overrides the default rhel-coreos base image.
+
+##### User workflow in layering model
```
The next few sections cover image stream based flows. Maybe such sections should be explicitly moved to a future section? Since I know what the plan for phase 0 is, I can read or understand those sections but someone not so close to the team may have a hard time understanding what's for now and what's for later.

sinnykumari replied:
done

sinnykumari commented:
Thanks Mrunal for the feedback! For better readability, it makes sense to keep enhancement focus on phase 0. I have updated enhancement focused on phase 0, moved remaining phases into a section `Future Phases`  and incorporated other review comments. PTAL updated EP, thanks!

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+
+**Note:** Rest of the enhancement proposal will focus on. For other phases, details can be found in [Future Phases](#future-phases) section.
+
+### Workflow Description
```
Can this be moved up under Phase 0?

sinnykumari replied:
As per enhancement template https://github.com/openshift/enhancements/blob/master/guidelines/enhancement_template.md#workflow-description  is a separate section inside proposal.

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+MCO will allow `osImageURL` field to be overridden via the MachineConfig.
+
+### Implementation details
+
```
Can we add a Phase 0 header here to indicate we are covering Phase 0 impl details?

sinnykumari replied:
It is not possible to add neatly just Phase0 header multiple times in the enhancement because linter throws error `MD024/no-duplicate-heading/no-duplicate-header Multiple headings with the same content [Context: "#### Phase 0"]`

So, I added a header `Phase 0 implementation`.  Also, I have explicitly callout in the note (at the end of Development Phases) that that rest of the implementation is about phase 0, so hopefully this shouldn't create any more confusion.

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+
+We need to preserve support for [extensions](https://github.com/openshift/enhancements/blob/master/enhancements/rhcos/extensions.md). For example, `kernel-rt` support is key to many OpenShift use cases.
+
+Extensions move to a `rhel-coreos-extensions` container that has RPMs. However, this is an implementation detail of the pipeline/MCO and should not be used by customers directly - only indirectly via specifying `kernelType` in a MachineConfig or via `extensions`.
```
Can we add some details on how extensions will work under the hood?

yuqi-zhang commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+
+### Upgrade / Downgrade Strategy
+
+Since this change requires latest rpm-ostree version. We will backport required rpm-ostree version into latest OCP 4.11.z release. Any older cluster upgrade would go through that 4.11.z release pointed via CVO upgrade graph.
```
Side note: this only affects the upgrade itself (need to pull new image to rebase to), but won't otherwise affect bootimage  issues which is tracked separately, right?

sinnykumari replied:
right. Later is covered in `Handling booting old nodes` section.

yuqi-zhang commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+Options:
+
+- Double reboot; but we'd still need to ship the old image format in addition to new
+  And really the only sane way to ship both is to generate the old from the new; we
```
minor nit: this seems to be the only place where we line break

yuqi-zhang commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+  - When cluster admin is ready to upgrade OS, they build new image derived from latest default RHEL CoreOS image supplied in release image and provides it via MachineConfig.
+  - At any point, the cluster can be switched back to default OS Image by deleting the custom osImageURL supplied via MachineConfig.
+
+#### Handling booting old nodes
```
I feel like this is important enough to warrant an explicit call out in the phase 0 plans, to keep the ability of booting old nodes once on the new format

sinnykumari replied:
umm, this whole section is talking mainly about phase 0. For even more clarity, will be adding phase 0 header explicitly in implementation details.

yuqi-zhang commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+1. The MCO builds a container image for each `MachineConfigPool`, using the new base image
+
+1. Each machineconfig pool will support a `custom-external-coreos` imagestream for pulling externally built images
+1.MCD continues to perform drains and reboots,, but for the most part does not change files in `/etc` or install packages per node
```
Nit: when rendered this isn't a separate bullet point since its missing a space after the `1.`. Also two commas

yuqi-zhang commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+
+We cannot just drop `MachineConfig` as an interface to node configuration. Hence, the MCO will be responsible for starting new builds on upgrades or when new machine config content is rendered.
+
+For most configuration, instead of having the MCD write files on each node, it will be added into the image build run on the cluster.
```
Implicitly, the MCD would still be able to do both to support RHEL nodes, right?

sinnykumari replied:
yes. that is explicitly mentioned in Non-Goals section

yuqi-zhang commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+
+Ignition will continue to handle the `disks` and `filesystem` sections - for example, LUKS will continue to be applied as it has been today.
+
+However, see below.
```
Are you meaning this in the sense that `However, there are some caveats`?

sinnykumari replied:
right. Colin can correct me if it meant something else.

yuqi-zhang reviewed (commented):
Did another pass and overall looks pretty good! Just some nits/question, doesn't have to block merging:

sinnykumari commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+They use this as part of a `Dockerfile` build that overrides the kernel, and just as above in the 3rd party scanner case, they find the digested pull spec and override via a `MachineConfig` with custom `osImageURL`.
+
+Later, a fixed kernel is released in the official `registry.redhat.io/rhel-coreos-8:4.12.8` image. The customer (and support) can easily see the kernel version in that image, and know with confidence that they can upgrade to a default cluster level of `4.12.8`.
+Specifically, the customer uses `oc adm upgrade` (or the GUI), and after that is complete they remove the `MachineConfig` object, which removes all the OS level overrides.
```
right, since user has overridden the default OS. They have better knowledge when is the right time to switch to default OS.

sinnykumari commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,730 @@
+Later, a fixed kernel is released in the official `registry.redhat.io/rhel-coreos-8:4.12.8` image. The customer (and support) can easily see the kernel version in that image, and know with confidence that they can upgrade to a default cluster level of `4.12.8`.
+Specifically, the customer uses `oc adm upgrade` (or the GUI), and after that is complete they remove the `MachineConfig` object, which removes all the OS level overrides.
+
+#### Externally built image
```
I think in cluster build usecase is relevant when we move MCO generated configs to the layering model.

sinnykumari commented:
@mrunalp  PTAL Updated enhancement with follow-up review comments

yuqi-zhang reviewed (approved):
Generally lgtm! Thanks a ton for working on this @sinnykumari !

aravindhp reviewed (commented):
Phase 0 looks clear though it is not clear what are the user stories that are resolved by it. 

PS: I am not a huge fan of the amount of future phases and plans called out here. It makes the enhancement really dense but I am sure that has been discussed and has been done with a reason. It also takes away from what is going to be delivered in the next release.

sinnykumari commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,738 @@
+#### Phase 3
+- oc command is augmented with sub-commands to provide syntactic sugar and a more polished synchronous user experience for cluster admins around staging a new OS image, watching preflight checks pass, and rolling it out. The web UI gains an equivalent UX flow.
+
+**Note:** Rest of the enhancement proposal will focus on. For other phases, details can be found in [Future Phases](#future-phases) section.
```
ah mentioning phase 0

sinnykumari commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,738 @@
+
+Moving to layering model requires some architectural changes in the MCO. Changes are also required in RHEL CoreOS base image format being shipped and the tooling around it. To rollout changes safely and minimize risk, we will take incremental development approach. Incremental development will also help us to get early feedback and iterate over overall design.
+
+### Development Phases
```
User stories are the end goals of what we would like to achieve from the whole implementation in supported way. Technically we would be able to achieve all these user stories upto certain extent by building custom image(derived from base RHCOS) and overriding osImageURL(phase 0) . But that won't be the ideal way we would expect users to use it and that's where multiple phases come into picture. 
We are trying to finalize phase 0. Remaining phases may change based on the feddback and experience we receive from pashe 0.

sinnykumari replied:
included in the notes.

sinnykumari commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,738 @@
+
+- While the base CoreOS layer/ostree-container tools will be usable outside of OpenShift, this enhancement does not cover or propose any in-cluster functionality for exporting the forward-generated image outside of an OpenShift cluster. In other words, it is not intended to be booted (`$ rpm-ostree rebase <image>`) from outside of a cluster.
+- This proposal does not cover generating updated "bootimages"; see https://github.com/openshift/enhancements/pull/201 and https://github.com/coreos/fedora-coreos-tracker/issues/1151.
+- Don't change existing workflow for RHEL worker nodes.
```
goo catch, fixed

sinnykumari commented:
> Phase 0 looks clear though it is not clear what are the user stories that are resolved by it.

Thanks for reviewing the EP!
About user stories, replied in https://github.com/openshift/enhancements/pull/1150#discussion_r947873928

> PS: I am not a huge fan of the amount of future phases and plans called out here. It makes the enhancement really dense but I am sure that has been discussed and has been done with a reason. It also takes away from what is going to be delivered in the next release.

Agree.
Little bit of context, initially we wanted to build everything altogether and that's where our lot of initial spikes and discussion happened which is reflected in initial commits in this PR (https://github.com/openshift/enhancements/pull/1150/commits/eb50da52055b251439a3515468f477237c0be03c) .
Later on, we realized to make it a bit simpler and take phased approach which will allow us to achieve the end result incrementally and with less risks involved.

So, things that doesn't fit into phase 0 has been moved to future plan which would be pulled in later into upcoming phases when as we have successfully implemented phase 0.

We can sync-up offline if more context is needed :)

sinnykumari commented:
Thank you all for reviewing this EP!
@mrunalp All the recent review comments has been addressed and as per offline discussion have squashed all commits into single commit. Should be ready for final review and hopefully we can get this merged :)

mrunalp commented on enhancements/ocp-coreos-layering/ocp-coreos-layering.md:
```diff
@@ -0,0 +1,741 @@
+
+ The goal for 4.12 is to have a supported path for users to apply hotfix (GA) into base RHEL CoreOS and ability to try out 3rd party software (Tech Preview).
+
+#### Dev Preview -> Tech Preview
```
Can we clean up this section since we are going straight to GA for hotfix?

sinnykumari replied:
done, hopefully this looks good now.
